Feat/gh actions

andthum opened the pull request:
Hopefully resolves failing GitHub Actions and documentation building

Changes proposed in this pull request:
  * Changed the requirements.txt files to be less strict regarding the required package versions
  * Changed the required Python version to now include Python 3.8.*
  * Formatted setup.py with black

@ezavod

ezavod commented:
This should solve most problems. You can try and add `python -m pip install numpy` in the github workflow right before the requirements. This should solve the last remaining problem (but not sure if this is an optimal solution).

What's the purpose of 'requirements_python3.8.5.txt'?

andthum commented:
I see requirements_python3.8.5.txt as a safety backup. With the dependencies and versions listed there, MDTools should be installed successfully using Python 3.8.5. So if at any point we were unable to install MDTools anymore due to version conflicts, we could can come back to requirements_python3.8.5.txt and install MDTools with the requirements listed there.

ezavod commented:
Ok, that makes sense. Technically we would have needed to just downgrade these versions:
```patch
-gsd==2.4.2
+gsd==2.4.1
-h5py==3.2.1
+h5py==3.1.0
-numpy==1.20.2
+numpy==1.19.5
-scipy==1.6.3
+scipy==1.5.4
```
But having too strict version requirements would mean that one would have to continuously mess with versions whenever something changes upstream.

ezavod commented on setup.py:
```diff
@@ -52,89 +52,94 @@ def get_varstring(fname="mdtools/version.py", varname="__version__"):
-            'Feature Requests': "TODO",
+            "Source": "https://github.com/andthum/mdtools",
+            "Documentation": "https://mdtools.readthedocs.io/en/latest/",
+            "Issue Tracker": "TODO",
+            "Feature Requests": "TODO",
```
I think one can link here `https://github.com/andthum/mdtools/issues` already instead of TODO.

andthum replied:
Good point.

ezavod commented on setup.py:
```diff
@@ -52,89 +52,94 @@ def get_varstring(fname="mdtools/version.py", varname="__version__"):
-                          "matplotlib >=3.2, <3.3",
-                          "MDAnalysis >=1.0, <2.0",
-                          "pyemma >=2.5, <3.0"]
+        python_requires=">=3.6",
```
Why drop the upper bound? In `classifiers` only 3.6, 3.7, 3.8 are listed.

andthum replied:
You are right, it should be `<3.9`, because MDAnalysis requires Python <3.9. I must have messed something in the commit.

ezavod commented on .github/workflows/tests.yml:
```diff
@@ -16,7 +16,7 @@ jobs:
       fail-fast: true
       matrix:
-        python-version: [3.6, 3.7, 3.8]
+        python-version: [3.7, 3.8]
```
I think we can include 3.6 if we prefetch numpy in the following section:
```patch
python -m pip install --upgrade pip
python -m pip install flake8 flake8-docstrings pytest
+ python -m pip install numpy
if [ -f requirements.txt ]; then pip install -r requirements.txt; fi
```

andthum replied:
Maybe replacing
```patch
- numpy >=1.18, <2.0
+ numpy >=1.17, <2.0
```
in requirements.txt will suffice?

ezavod replied:
I think, since we cannot rely on the exact order in requirements, we have to install numpy first. This might not necessarily bound by a specific version because the requirements entry is read eventually. But it prevents some third party dependency to pull incompatible versions.

andthum replied:
Seems like this does the trick.

ezavod reviewed (approved):
Looks good, thanks for fixing the problem I introduced :-)
Would you like to squash the redundant commits before merging?

andthum commented:
Well, the errors were there before, you just spotted them with your CI workflow ;)

I'll take a look how to squash individual commits. Currently, I've no idea how to do that.
Generally, I do not think that we should require contributors to squash their commits (see also our Discussion #18). But of course contributors can squash, if they like, to have a cleaner history.

andthum commented:
I reordered and squashed the redundant commits. I also added a badge to the README and the docs indicating whether the tests (GitHub Actions) pass or fail.

Probably, it's not the best idea to rewrite the history of a PR which was already commented and discussed, because then the discussion seems very odd. So, relating to the discussion #18 we had, we should encourage contributors to clean up their commits *before* pushing, although we shouldn't (and can't) require this. In this special case where we want to test GitHub Actions, cleaning the history beforehand is of course not possible at all.

ezavod reviewed (approved):
Looks good, but the last commit "Merge ..." does not seem to have any function and is empty, isn't it?
Generally all in-code discussion is permanent and just marked as outdated and the rough changes can be viewed by looking at the force-pushed diffs. This means if changes are rather small keeping track of what's going on should not be too difficult. But I also agree that one should aim at PRs that are as clean as possible when submitting stuff.

andthum commented:
> [...] the last commit "Merge ..." does not seem to have any function and is empty, isn't it?

Very good point. GitHub said, that I have to update this feature branch. There was an "Update" button and after I pressed it, GitHub apparently merged the `main` branch *into* the `feat/gh_actions` branch. I have do undo the last commit, because right now, we have no `main` branch anymore (see https://github.com/andthum/mdtools/network).

Another bad thing is that the `feat/gh_actions` branch now branches off from `feat/mdtools_logo_for_GitHub`. I must have messed up some things, when rewriting the history of `feat/gh_actions`. I will fix this before merging.

andthum commented:
Ok, I removed the last commit that merged merged the `main` branch into the `feat/gh_actions` branch and rebased the `feat/gh_actions` branch onto the `main` branch (instead of the `cleanup` branch.

I need approval before merging again (if this is annoying, we can allow merging without approved review).

ezavod commented:
All the buttons seem to be very weird. On the CLI it's as simple as typing rebase :-D
